x/supplier/keeper/msg_server_stake_supplier.go

```diff
@@ -296,26 +296,48 @@ func (k Keeper) updateSupplier(
 	currentHeight := sdkCtx.BlockHeight()
 	nextSessionStartHeight := sharedtypes.GetNextSessionStartHeight(&sharedParams, currentHeight)
 
-	// If new service configs were provided, mark all the supplier's service
-	// configurations as to be deactivated at the start of the next session.
-	if len(msg.Services) > 0 {
-		for _, oldServiceConfigUpdate := range supplier.ServiceConfigHistory {
-			oldServiceConfigUpdate.DeactivationHeight = nextSessionStartHeight
-		}
-	}
+	updatedServiceConfigHistory := make([]*sharedtypes.ServiceConfigUpdate, 0)
+	updatedServices := make(map[string]struct{})
 
-	// Initialize the supplier's service configurations with the new ones from the msg.
-	// These will take effect at the start of the next session.
+	// Step 1: Add all new service configurations from the message
+	// - These configs will activate at the start of the next session
+	// - Track service IDs to identify which old inactive configs need replacement
 	for _, newServiceConfig := range msg.Services {
 		newServiceConfigUpdate := &sharedtypes.ServiceConfigUpdate{
 			OperatorAddress: msg.OperatorAddress,
 			Service:         newServiceConfig,
 			// The effective block height is the start of the next session.
 			ActivationHeight: nextSessionStartHeight,
 		}
-		supplier.ServiceConfigHistory = append(supplier.ServiceConfigHistory, newServiceConfigUpdate)
+		updatedServiceConfigHistory = append(updatedServiceConfigHistory, newServiceConfigUpdate)
+		updatedServices[newServiceConfig.ServiceId] = struct{}{}
+	}
+
```

Comment:
This is optional, but I feel like it's clearer (not necessarily safer) with regard to determinsim.

1. Populate the map above
2. Create a list here and sort it by keys
3. Use it below
4. Add comments w.r.t to the "why"

<img width="994" height="463" alt="Screenshot 2025-10-07 at 4 02 32 PM" src="https://github.com/user-attachments/assets/c1652bf8-0382-4aac-80d1-b21e60b3bc20" />